core/deadline.go

```diff
@@ -34,18 +36,135 @@ type slotTimeProvider interface {
 	eth2client.SlotDurationProvider
 }
 
-// Deadliner provides duty Deadline functionality. The C method isn’t thread safe and may only be used by a single goroutine. So, multiple
-// instances are required for different components and use cases.
+// Deadliner provides duty Deadline functionality. The C method isn’t thread safe and
+// may only be used by a single goroutine. So, multiple instances are required
+// for different components and use cases.
 type Deadliner interface {
-	// Add adds a duty to be notified of the Deadline via C. Note that duties will be deduplicated and only a single duty will be provided via C.
+	// Add adds a duty to be notified of the Deadline via C.
+	// Note that duties will be deduplicated and only a single duty will be provided via C.
 	Add(duty Duty)
 
 	// C returns the same read channel every time and contains deadlined duties.
 	// It should only be called by a single goroutine.
 	C() <-chan Duty
 }
 
+// Deadline implements the Deadliner interface.
+type Deadline struct {
+	mu           sync.Mutex
+	genesisTime  time.Time
+	slotDuration time.Duration
+
+	dutyChan     chan Duty
+	deadlineChan chan Duty
+
+	// duties represents a set since duty deduplication is required.
+	duties map[Duty]bool
+}
+
+// NewDeadliner returns a new instance of Deadline.
+// It runs a goroutine which is responsible for reading and storing duties,
+// and sending the deadlined duty to receiver's deadlineChan.
+func NewDeadliner(ctx context.Context, eth2Cl slotTimeProvider) (*Deadline, error) {
+	genesis, err := eth2Cl.GenesisTime(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	duration, err := eth2Cl.SlotDuration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Deadline{
+		genesisTime:  genesis,
+		slotDuration: duration,
+		dutyChan:     make(chan Duty),
+		deadlineChan: make(chan Duty),
+		duties:       make(map[Duty]bool),
+	}
+
+	go func() {
+		var (
+			deadlineTimer <-chan time.Time
+			minDuty       Duty
+		)
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case duty := <-d.dutyChan:
+				if minDuty.Type == DutyUnknown {
+					// Initialise minDuty and deadlineTimer
+					minDuty = duty
+					deadlineTimer = time.After(time.Until(d.deadlineTime(minDuty)))
+				}
+				d.duties[duty] = true
+			case <-deadlineTimer:
```

Comment:
Note that this will block forever when DutyExit is selected as minDuty....

Reply:
So you basiaclly need to update the current deadline timer if a duty is received with an earlier deadline. So if you are currently on super long DutyExit deadline, and you receive DutyAttestation for the next slot (which has earlier deadline), then you need to replace that deadline with the new duty.

suggest:
```
var (
  currentTimer     *time.Timer
  currentDuty       Duty
  currentDeadline time.Time
)

    case duty := <-d.dutyChan:
       deadline := deadlineFunc(duty)
       if currDeadline.After(deadline) {
           // Replace with newer deadline
           currentTimer.Stop()
           currentDuty = duty
	   currentTimer = time.NewTimer(time.Until(d.deadlineFunc(duty)))
	   currentDeadline = deadline
       }
```